### PR TITLE
[Datasets] Add `ImageFolderDatasource`

### LINKED ISSUE #23977: [Datasets] [Feature] Add datasource for canned ML datasets and imagery data sources
Being able to quickly work with familiar canned datasets exposed by ML frameworks (e.g. CIFAR10, MNIST, ImageNet, etc.) is great for a "getting started" flow, since the user doesn't have to worry about finding a dataset to try out and the user is familiar with these common ML datasets. We should add a Datasets datasource for some of these canned datasets to improve our "getting started" UX.

### Requirements

1. [P0] Expose a generic Torch dataset wrapper that downloads the entire dataset onto a single node and builds a single block. This should provide an easy "getting started" path for smaller canned datasets.
2. [P0] Do the same for TensorFlow datasets.
3. [P0] Validate two common image loading workflows for larger-scale (multi-node, sharded) datasets:
  a. **Image-only dataset:** Take a list of image URLs, map a download function over them that (i) parses the class from the path, and (ii) downloads the image. To test this, add one or two of the larger canned datasets (e.g. ImageNet is multiple TB) to a public S3 bucket (or find a large one that’s already hosted). `ray.data.from_items(s3_image_urls).map(download_fn)`
  b. **Images-in-tables dataset:** Read a tabular (Parquet, CSV) dataset that contains an image URL column, map a download function over the blocks that takes the image URL for each row and adds the image to a new column. `ray.data.read_parquet(tabular_ds_url).map_batches(...)`
4. [P1] If 3(a) proves to be bad UX, create a custom imagery-specific datasource that reads the data in parallel from said bucket. This should demonstrate some of the scalable parallel I/O benefits of Datasets and will be a nice “how would I scale this to larger datasets” progression.

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

Popular datasets like ImageNet and Tiny ImageNet are arranged in a specific layout like this:

```
root/dog/xxx.png
root/dog/xxy.png
root/dog/[...]/xxz.png

root/cat/123.png
root/cat/nsdf3.png
root/cat/[...]/asd932_.png
```

This PR adds a datasource that reads such datasets.

## Related issue number

Closes #23977 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/impl/pandas_block.py
Comment (bveeramani) [inline python/ray/data/impl/pandas_block.py:57]: This change is necessary so that users can do

`ds.take(...)`

on a `Dataset` read by `ImageFolderDatasource`.

File: python/ray/data/datasource/file_meta_provider.py
Comment (bveeramani) [inline python/ray/data/datasource/file_meta_provider.py:203]: I've removed this warning because I think it'll lead to a better UX for the user.

If the user calls `ray.data.read_datasource(ImageFolderDatasource(), ...)`, a warning is raised:

> (_prepare_read pid=43380) 2022-05-11 02:11:33,272       WARNING [file_meta_provider.py:199](http://file_meta_provider.py:199/) -- Skipping expansion of 2 path(s). If your paths contain directories or if file size collection is required, try rerunning this read with `meta_provider=DefaultFileMetadataProvider()`.

This may confuse the user because:
1. It has nothing to do with the user's problem
2. It leaks an implementation detail
3. It can't be resolved

Because `ImageFolderDatasource` is user-facing and `FastFileMetadataProvider` is a developer API, I think it's overall better to remove the warning.

Comment (ericl) [reply python/ray/data/datasource/file_meta_provider.py:203]: Hey please split this change into a separate PR and raise it in the data team for discussion.

File: doc/source/data/package-ref.rst
Comment (matthewdeng) [inline doc/source/data/package-ref.rst:138]:
```suggestion
.. autoclass:: ray.data.datasource.ImageFolderDatasource
    :members:
```

File: python/ray/data/datasource/image_folder_datasource.py
Comment (matthewdeng) [inline python/ray/data/datasource/image_folder_datasource.py:75]: Make sure it's not empty either?
```suggestion
        if len(paths) != 1:
```

File: python/ray/data/datasource/image_folder_datasource.py
Comment (matthewdeng) [inline python/ray/data/datasource/image_folder_datasource.py:37]: nit: Call out that `dog` and `cat` are labels?

File: python/ray/data/datasource/image_folder_datasource.py
Comment (matthewdeng) [inline python/ray/data/datasource/image_folder_datasource.py:98]: Are there any scalability concerns expanding all the file paths into memory?

File: doc/source/data/package-ref.rst
Comment (bveeramani) [reply doc/source/data/package-ref.rst:138]: `:members:` would only add `ImageFolderDatasource.prepare_read`. IMO it's better to leave it out

> prepare_read(parallelism: int, paths: Union[str, List[str]], filesystem: Optional[pyarrow.fs.FileSystem] = None, columns: Optional[List[str]] = None, schema: Optional[Union[type, pyarrow.lib.Schema]] = None, meta_provider: ray.data.datasource.file_meta_provider.ParquetMetadataProvider = <ray.data.datasource.file_meta_provider.DefaultParquetMetadataProvider object>, _block_udf: Optional[Callable[[Union[List[ray.data.block.T], pyarrow.Table, pandas.DataFrame, bytes]], Union[List[ray.data.block.T], pyarrow.Table, pandas.DataFrame, bytes]]] = None, **reader_args)→ List[[ray.data.datasource.datasource.ReadTask](https://docs.ray.io/en/master/data/package-ref.html#ray.data.ReadTask)][[source]](https://docs.ray.io/en/master/_modules/ray/data/datasource/parquet_datasource.html#ParquetDatasource.prepare_read)
Creates and returns read tasks for a Parquet file-based datasource.

File: python/ray/data/datasource/file_meta_provider.py
Comment (bveeramani) [reply python/ray/data/datasource/file_meta_provider.py]: Done. See #24909

File: python/ray/data/datasource/image_folder_datasource.py
Comment (bveeramani) [reply python/ray/data/datasource/image_folder_datasource.py:37]: Do you feel like it's not obvious that `dog` and `cat` are the labels? I feel like based on convention it should be clear, but maybe I'm suffering from the curse of knowledge.

File: python/ray/data/datasource/image_folder_datasource.py
Comment (bveeramani) [reply python/ray/data/datasource/image_folder_datasource.py:100]: > At a high level, I think this PR can be considerably simplified if you wait for @clarkzinzow 's PR adding file-extension filters to merge: https://github.com/ray-project/ray/pull/24822

I agree. Since this PR isn't particularly time sensitive, I'll wait for #24822 to get merged.

**Comment by clarkzinzow** [comment]
@ericl This datasource is a bit opinionated about the layout, e.g. expecting it to be directory-partitioned on the class label. For this to be a generic `read_images()` API, we'd need to tweak this so those opinions are less strong, or are at least overridable defaults.

**Comment by ericl** [comment]
> @ericl This datasource is a bit opinionated about the layout, e.g. expecting it to be directory-partitioned on the class label. For this to be a generic read_images() API, we'd need to tweak this so those opinions are less strong, or are at least overridable defaults.

Hmm, what if we followed the read_binary_files convention and included the path by default? We could allow the user to specify a custom `label_function(path: str) -> str` that if specified is used to generate a `label` column in place of the `path` column.

**Comment by ericl** [comment]
So we could call it `read_image_files`, to be similar to `read_binary_files`. A couple points were raised also in other discussions:
- We should return the data in Arrow format (Tensor image column, other metadata columns), instead of simple python objects like read_binary_files() does.
- The folder thing can follow existing partitioning convention from pyarrow (https://arrow.apache.org/docs/python/generated/pyarrow.dataset.partitioning.html). Basically, there's nothing special about image folders, they can be handled the same as folders of other files like parquet.

Can you work with @clarkzinzow to look at partitioning support? I think it could be as simple as supporting: `ray.data.read_image_files(path, partitioning=pyarrow.dataset.partitioning(field_names=["label"]))` (we can only support directory partitioning).

> @ericl One of the users we spoke to mentioned that he'd seen image data stored like this (in a CSV or Parquet or something):

For this case, perhaps we should have a image decoding utility function they can map_batches()...

File: python/ray/data/datasource/image_folder_datasource.py
Comment (amogkam) [inline python/ray/data/datasource/image_folder_datasource.py:121]: We have to use TensorArray/TensorArrayElement here right? Otherwise I don't think this will work with Predictors during tensor conversion.

cc @clarkzinzow

**Comment by amogkam** [comment]
Can we make sure this works with the CUJ before merging this in 

```python
s3_url = "s3://air-cuj-imagenet-20gb"

import numpy as np
import pandas as pd
from io import BytesIO
from PIL import Image
import ray
from ray.air.util.tensor_extensions.pandas import TensorArray
from ray.train.torch import to_air_checkpoint, TorchPredictor
from ray.train.batch_predictor import BatchPredictor

from torchvision import transforms
from torchvision.models import resnet18

from ray.data.preprocessors import BatchMapper

def convert_to_pandas(byte_item_list):
    images = [Image.open(BytesIO(byte_item)).convert('RGB') for byte_item in byte_item_list]
    images = [np.array(image) for image in images]

    return pd.DataFrame({"image": TensorArray(images)})

# What is parallelism magic number?
dataset = ray.data.read_binary_files(paths=s3_url)
dataset = dataset.map_batches(convert_to_pandas)

def preprocess(df):
    preprocess = transforms.Compose([
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
    ])
    df["image"] = df["image"].map(preprocess)
    return df

preprocessor = BatchMapper(preprocess)

model = resnet18(pretrained=True)

# # TODO: to_air_checkpoint should accept a model or a model state dict
ckpt = to_air_checkpoint(model=model, preprocessor=preprocessor)

predictor = BatchPredictor.from_checkpoint(ckpt, TorchPredictor)
predictor.predict(dataset, num_gpus_per_worker=1)
```

When I had previously tried copying this code to use for the CUJ, it failed bc of some TensorArray/TensorArrayElement conversion

File: python/ray/data/datasource/image_folder_datasource.py
Comment (clarkzinzow) [reply python/ray/data/datasource/image_folder_datasource.py:121]: @amogkam Yep, correct!

Comment (bveeramani) [reply python/ray/data/datasource/image_folder_datasource.py:121]: > Can we make sure this works with the CUJ before merging this in

Yeah, will do

**Comment by bveeramani** [comment]
> Can we make sure this works with the CUJ before merging this in

Yeah, will do

**Comment by richardliaw** [comment]
For the CUJ that amog posted can we add it in as an example to test in ci?

File: python/ray/data/read_api.py
Comment (bveeramani) [reply python/ray/data/read_api.py:1078]: No strong opinion. 

If there's no consensus, I'll add the datasource now. We can always add a `read_*` API later.

Comment (richardliaw) [reply python/ray/data/read_api.py:1078]: Let's do that.

**Comment by bveeramani** [comment]
> For the CUJ that amog posted can we add it in as an example to test in ci?

No. There are issues with applying TorchVision transformations that are completely unrelated to this PR.

To make this code snippet work, you need to add several workarounds

```
def preprocess(df):
    preprocess = transforms.Compose([
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
    ])
    df["image"] = df["image"].map(preprocess)
    return df
```

```
def preprocess(df):
    preprocess = transforms.Compose([
        lambda ray_tensor: ray_tensor.to_numpy(),
        transforms.ToTensor(),
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
        lambda torch_tensor: torch_tensor.numpy().astype(np.float32)
    ])
    df["image"] = TensorArray([preprocess(image) for image in df["image"]])
    return df
```

File: python/ray/data/datasource/image_folder_datasource.py
Comment (bveeramani) [reply python/ray/data/datasource/image_folder_datasource.py:63]: > Is this mapping available as metadata in the Dataset anywhere?

I'm not sure if I totally understand what you mean. Can you give me more information?

**Comment by richardliaw** [comment]
@bveeramani sorry, what I mean is we should have an end-to-end test such as Amog's example (with whatever modifications you want to make).

Can you please do that before merging?

File: python/ray/data/read_api.py
Comment (clarkzinzow) [reply python/ray/data/read_api.py:1078]: It might be worth adding this base API for the sake of discoverability.

**Comment by bveeramani** [comment]
> Can you please do that before merging?

Added and E2E test. Wasn't sure what to test other than that there are no errors

File: python/ray/data/datasource/image_folder_datasource.py
Comment (richardliaw) [inline python/ray/data/datasource/image_folder_datasource.py:41]: can you still raise ImportError here?

The problem here is that there is a missing dependency right?

Comment (bveeramani) [reply python/ray/data/datasource/image_folder_datasource.py:41]: Yeah

File: python/ray/data/read_api.py
Comment (bveeramani) [reply python/ray/data/read_api.py:1078]: Relevant conversation on Slack:

**Balaji**
Hey, do you guys have any strong opinions about whether ImageFolderDatasource should have a corresponding read_* API?

**Clark**
I think it might be worth having, let’s discuss on the PR thread though: https://github.com/ray-project/ray/pull/24641/files#r921790567

**ekl**
I thought the issue was we were punting proper partitioning suppotr, so it doesn't make sense to have it first-class for now?

**Balaji**
By proper partitioning support, do you mean supporting arbitrary partitions, rather than assuming we partition class by the first directory after the root?

**ekl**
that's right (following standard pyarrow partitioning conventions)

**Balaji**
I think we'll be able to add proper partitioning support without making breaking changes to the read_* API

Since the functionality is useful without full partitioning support, maybe we should add a top-level API for the sake of discoverability (as Clark suggested on the thread)? (edited) 

**ekl**
I'm ok with that.

Comment (richardliaw) [reply python/ray/data/read_api.py:1078]: Please, let's go with the consensus that was agreed on slack, just to avoid frustrations for others in the discussion.

Opening a new PR after with the read_* API shouldn't be too hard :)

**Review by ericl — APPROVED** [review]

File: python/ray/data/datasource/image_folder_datasource.py
Comment (matthewdeng) [reply python/ray/data/datasource/image_folder_datasource.py]: Ah not an issue anymore since we dropped `target` column. But the original intent was to ask how the user would access `label_to_target` from the created Dataset.

**Review by clarkzinzow — APPROVED** [review]
LGTM, great work!

File: python/ray/data/tests/test_dataset_formats.py
Comment (matthewdeng) [inline python/ray/data/tests/test_dataset_formats.py:2870]: Move these tests above line 2735 as they use the `ray_start_regular_shared` fixture.

Comment (richardliaw) [reply python/ray/data/tests/test_dataset_formats.py]: fixed

File: python/ray/data/datasource/image_folder_datasource.py
Comment (jiaodong) [inline python/ray/data/datasource/image_folder_datasource.py:127]: do we have any docs / past discussion about this part? Basically we're assuming we get the label based on user file path, which has to be structured in certain way in order to get the correct one without knobs needed to pass in custom label file or join ?

For example, if i read a s3 bucket with filenames of "dog.jpg", "dog_2.jpg" my dataloader will end up getting these string values by default.

Comment (bveeramani) [reply python/ray/data/datasource/image_folder_datasource.py:127]: > Basically we're assuming we get the label based on user file path, which has to be structured in certain way in order to get the correct one without knobs needed to pass in custom label file or join ?

Yeah, that's right. The datasource assumes that the layout is structured in the same way as ImageNet. The functionality of the datasource is based on that of [TorchVision's ImageFolder](https://pytorch.org/vision/stable/generated/torchvision.datasets.ImageFolder.html).

> For example, if i read a s3 bucket with filenames of "dog.jpg", "dog_2.jpg" my dataloader will end up getting these string values by default.

Yeah, you're right. We don't validate that the label corresponds to a directory. In this case, we could raise an error stating that the folder isn't structured correctly. 

Alternatively, if images aren't stored in a directory, we could set the label to `None`.

```
If images aren't stored in a sub-directory, then the image's label will be set to `None`.

.. code-block::

    root/dog/xxx.png  # Label is 'dog'
    root/123.jpg.     # Label is `None`
```

**Pull request merged**
